multicluster/controllers/multicluster/core/leader_elector.go

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2021 Antrea Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"crypto/rand"
+	"math/big"
+
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/multicluster/controllers/multicluster/common"
+)
+
+/**
+ * Lets consider a ClusterSet with 2 leader clusters and 1 member cluster.
+ *  Leader1        Leader2
+ *       \         /
+ *        \       /
+ *         -------
+ *            ^
+ *            |
+ *         Member1
+ * The member cluster connects to both leaders. It writes MemberClusterAnnounce
+ * and other resources into CommonArea of both leaders. However, resources
+ * are imported only from an elected-leader cluster.
+ *
+ * This interface performs the election to pick an elected-leader
+ * from the list of leader clusters in the ClusterSet spec.
+ *
+ * Leader election picks an elected-leader among "connected" leaders
+ * by choosing one randomly.
+ * The member cluster periodically writes MemberClusterAnnounce into every leader
+ * cluster and is considered "connected" if it can successfully write so.
+ * After leader election is done, the result is also used to update the
+ * spec of MemberClusterAnnounce during the next periodic write into the
+ * leader's CommonArea, so the leader knows it is the elected-leader of this
+ * member (mainly for visibility and reporting).
+ */
+
+type LeaderElector interface {
+	// RunLeaderElection runs leader election. It is invoked periodically on a timer.
+	RunLeaderElection()
+}
+
+func (r *remoteCommonAreaManager) RunLeaderElection() {
+	if !r.needElection {
+		if r.electedLeaderCluster != nil {
+			if !r.electedLeaderCluster.IsConnected() {
+				klog.InfoS("Disconnected leader", "Cluster",
+					r.electedLeaderCluster.GetClusterID())
+				r.setElectedLeader(nil)
+			}
+		}
+		if r.electedLeaderCluster == nil {
+			// Do we have any member that is connected?
+			for _, cluster := range r.remoteCommonAreas {
```

Comment:
It helps me if we have a comment to say remoteCommonAreas is accessed only in a single goroutine (so locking is not necessary).

Reply:
I have the comment on the field
`	// remoteCommonAreas is a map of RemoteCommonArea.
	// It is accessed only from the event-loop in a single goroutine context.
	remoteCommonAreas map[common.ClusterID]RemoteCommonArea`

I have comments on fields that require locking and why need it, so for others it implies locking is not needed. Do you think that is good or should I elaborate anything?

Reply:
Ok. Comments on the fields should be good enough.